mw_basic_auth.go

```diff
@@ -117,6 +111,66 @@ func (k *BasicAuthKeyIsValid) ProcessRequest(w http.ResponseWriter, r *http.Requ
 		ctxSetAuthToken(r, keyName)
 	}
 
-	// Request is valid, carry on
 	return nil, http.StatusOK
 }
+
+func (k *BasicAuthKeyIsValid) handleAuthFail(w http.ResponseWriter, r *http.Request, token string) (error, int) {
+
+	// Fire Authfailed Event
+	AuthFailed(k, r, token)
+
+	// Report in health check
+	reportHealthValue(k.Spec, KeyFailure, "-1")
+
+	return k.requestForBasicAuth(w, "User not authorised")
+}
+
+func (k *BasicAuthKeyIsValid) compareHashAndPassword(hash string, password string, logEntry *logrus.Entry) error {
+
+	cacheEnabled := !k.Spec.BasicAuth.DisableCaching
+	cacheDuration := time.Duration(60) * time.Second
+	if k.Spec.BasicAuth.CacheTTL > 0 {
+		cacheDuration = time.Duration(k.Spec.BasicAuth.CacheTTL) * time.Second
+	}
+
+	hasher := murmur3.New32()
+	passwordBytes := []byte(password)
+
+	doBcrypt := func(cacheEnabled bool, cacheDuration time.Duration, hashedPassword string, password []byte) error {
+		if err := bcrypt.CompareHashAndPassword([]byte(hash), password); err != nil {
```

Comment:
one confusing thing - shouldn't this be skipped if `cacheEnabled` is true?

Reply:
nop - because if not in cache, or hash doesn't match from cache, then it fails-over to bcrypt as src of truth. doBcrypt also populates the cache if cache enabled